t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+            local g = getmetatable(_G).__index
+
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    return
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:wait(1)
+            if ok then
+                ngx.print("wait")
+            end
+        ';
+    }
+
+    location /sem_post {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            ngx.sleep(0.001)
+            collectgarbage("collect")
+            local ok, err = sem:post()
+            if ok then
+                ngx.print("post")
+                ngx.exit(200)
+            else
+                ngx.exit(500)
+            end
+
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- no_error_log
+[error]
+[crit]
+
+
+
+=== TEST 2: basic semaphore wait post
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local ok, err = sem:wait(0.1)
+            ngx.say(tostring(ok).." "..tostring(err))
+        ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+false timeout
+--- no_error_log
+[error]
+
+
+
+=== TEST 3: basic semaphore wait not allow in body_filter_by_lua
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+        body_filter_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local ok, err = sem:wait(1)
+            ngx.log(ngx.ERR, err)
+        ';
+        return 200;
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+--- error_log
+API disabled in the context of (unknown)
+
+
+
+=== TEST 4: basic semaphore wait not allow in header_filter_by_lua
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+        header_filter_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            local ok, err = sem:wait(1)
+            ngx.log(ngx.ERR, err)
+        ';
+        return 200;
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+--- error_log
+API disabled in the context of header_filter_by_lua*
+
+
+
+=== TEST 5: basic semaphore wait not allow in init_worker_by_lua
+--- http_config eval: $::HttpConfigIntWorkerBy
+--- config
+    location /test {
+       content_by_lua '
```

Comment:
This complicated `content_by_lua` directive is useless for the test case. Better replace it with something trivial like `echo ok`.